src/libraries/System.Net.Security/src/System/Net/Security/SslStream.Implementation.cs

```diff
@@ -826,109 +933,62 @@ private async ValueTask<int> ReadAsyncInternal<TIOAdapter>(TIOAdapter adapter, M
-                        // This zero-byte read indicates their desire to trade off the extra cost of a zero-byte read
-                        // for reduced memory consumption when data is not immediately available.
-                        // So, we will issue our own zero-byte read against the underlying stream and defer buffer allocation
-                        // until data is actually available from the underlying stream.
-                        // Note that if the underlying stream does not supporting blocking on zero byte reads, then this will
-                        // complete immediately and won't save any memory, but will still function correctly.
-                        await adapter.ReadAsync(Memory<byte>.Empty).ConfigureAwait(false);
+                        // We either filled whole buffer or used all buffered frames.
+                        return processedLength;
                     }
 
-                    ResetReadBuffer();
-
-                    // Read the next frame header.
-                    if (_internalBufferCount < SecureChannel.ReadHeaderSize)
-                    {
-                        // We don't have enough bytes buffered, so issue an initial read to try to get enough.  This is
-                        // done in this method both to better consolidate error handling logic (the first read is the special
-                        // case that needs to differentiate reading 0 from > 0, and everything else needs to throw if it
-                        // doesn't read enough), and to minimize the chances that in the common case the FillBufferAsync
-                        // helper needs to yield and allocate a state machine.
-                        int readBytes = await adapter.ReadAsync(_internalBuffer.AsMemory(_internalBufferCount)).ConfigureAwait(false);
-                        if (readBytes == 0)
-                        {
-                            return 0;
-                        }
-
-                        _internalBufferCount += readBytes;
-                        if (_internalBufferCount < SecureChannel.ReadHeaderSize)
-                        {
-                            await FillBufferAsync(adapter, SecureChannel.ReadHeaderSize).ConfigureAwait(false);
-                        }
-                    }
-                    Debug.Assert(_internalBufferCount >= SecureChannel.ReadHeaderSize);
+                    buffer = buffer.Slice(processedLength);
+                }
 
-                    // Parse the frame header to determine the payload size (which includes the header size).
-                    int payloadBytes = GetFrameSize(_internalBuffer.AsSpan(_internalOffset));
-                    if (payloadBytes < 0)
-                    {
-                        throw new IOException(SR.net_frame_read_size);
-                    }
+                if (_receivedEOF && _internalBufferCount == 0)
```

Comment:
If I'm understanding the logic correctly, then if _receivedEOF is true, _internalBufferCount should always be 0. Right? Can we just assert that here rather than checking it?

Reply:
yes, I think you are right. I removed the second condition and I added Debug.Assert